C/services/notification-plugin-interfaces/python/python_plugin_interface.cpp

```diff
@@ -0,0 +1,657 @@
+	PyObject* pReturn = PyObject_CallFunction(pFunc,
+						  "Os",
+						  handle,
+						  assetValues.c_str());
+
+	Py_CLEAR(pFunc);
+
+	// Handle return
+	if (!pReturn)
+	{
+		Logger::getLogger()->error("Called python script method 'plugin_eval' : "
+					   "error while getting result object, plugin '%s'",
+					   sPluginName.c_str());
+		logErrorMessage();
+	}
+	else
+	{
+		// check bool abd return true or false
+		if (PyBool_Check(pReturn))
+		{
+			ret = PyObject_IsTrue(pReturn);
+		}
+	}
+	PyGILState_Release(state);
+
+	return ret;
+}
+
+/**
+ * Function to invoke 'plugin_reconfigure' function in python plugin
+ *
+ * @param    handle     The plugin handle from plugin_init_fn
+ * @param    config     The new configuration, as string
+ */
+void notification_plugin_reconfigure_fn(PLUGIN_HANDLE handle,
+					const std::string& config)
+{
+	if (!pModule)
+	{
+		Logger::getLogger()->fatal("plugin_handle: plugin_reconfigure(): "
+					   "pModule is NULL for plugin '%s'",
+					   sPluginName.c_str());
+		return;
+	}
+	if (!handle)
+	{
+		Logger::getLogger()->fatal("plugin_handle: plugin_reconfigure(): "
+					   "handle is NULL for plugin '%s'",
+					   sPluginName.c_str());
+		return;
+	}
+
+	std::mutex mtx;
+	lock_guard<mutex> guard(mtx);
+
+	PyGILState_STATE state = PyGILState_Ensure();
+
+	Logger::getLogger()->debug("plugin_handle: plugin_reconfigure(): "
+				   "pModule=%p, handle=%p, plugin '%s'",
+				   pModule,
+				   handle,
+				   sPluginName.c_str());
+
+	PyObject* pFunc;
+	// Fetch required method in loaded object
+	pFunc = PyObject_GetAttrString(pModule, "plugin_reconfigure");
+	if (!!pFunc)
+	{       
+		Logger::getLogger()->fatal("Cannot find method 'plugin_reconfigure' "
+					   "in loaded python module '%s'",
+					   sPluginName.c_str());
+	}
+
+	if (!pFunc || !PyCallable_Check(pFunc))
+	{
+		// Failure
+		if (PyErr_Occurred())
+		{
+			logErrorMessage();
+		}
+
+		Logger::getLogger()->fatal("Cannot call method 'plugin_reconfigure' "
+					   "in loaded python module '%s'",
+					   sPluginName.c_str());
+					   Py_CLEAR(pFunc);
+		PyGILState_Release(state);
+		return;
+	}
+
+	Logger::getLogger()->debug("plugin_reconfigure with %s", config.c_str());
+
+	// Call Python method passing an object and a C string
+	PyObject* pReturn = PyObject_CallFunction(pFunc,
+						  "Os",
+						  handle,
+						  config.c_str());
+
+	Py_CLEAR(pFunc);
+
+	// Handle returned data
+	if (!pReturn)
+	{
+		Logger::getLogger()->error("Called python script method plugin_reconfigure "
+					   ": error while getting result object, plugin '%s'",
+					   sPluginName.c_str());
+		logErrorMessage();
+		//*handle = NULL; // not sure if this should be treated as unrecoverable failure on python plugin side
+		PyGILState_Release(state);
+		return;
+	}
+	else
+	{
+		// Save curreent handle
+		PyObject* tmp = (PyObject*)handle;
+		// Create a new handle
+		handle = PyObject_Init(pReturn, Py_TYPE(pReturn));
```

Comment:
Why is there a need to call PyObject_Init on pReturn? pReturn is already a new reference returned by PyObject_CallFunction.

Reply:
I did that in order to avoid the Python valueError: "Circular reference detected"